vault/hcp_link/capabilities/api_capability/api_passthrough.go

```diff
@@ -65,7 +68,14 @@ func (p *APIPassThroughCapability) Start() error {
 	}
 
 	go func() {
-		err = p.scadaServer.Serve(listener)
+		if p.scadaServer.TLSConfig == nil {
+			err = p.scadaServer.Serve(listener)
+		} else {
+			// certFile and keyFile not required as server's
+			// TLSConfig.GetCertificate will be used
+			err = p.scadaServer.ServeTLS(listener, "", "")
```

Comment:
Do we need to call `ServeTLS` here? I think `ServeTLS` calls `Serve`. The only difference is that it uses the provided `cert` and `key` to initialize the TLS configuration before calling `Serve`.

Reply:
The only other difference is the function that's called to setup HTTP2. However, if we look at the underlying implementation, there's no real difference:

```
// setupHTTP2_ServeTLS conditionally configures HTTP/2 on
// srv and reports whether there was an error setting it up. If it is
// not configured for policy reasons, nil is returned.
func (srv *Server) setupHTTP2_ServeTLS() error {
	srv.nextProtoOnce.Do(srv.onceSetNextProtoDefaults)
	return srv.nextProtoErr
}

// setupHTTP2_Serve is called from (*Server).Serve and conditionally
// configures HTTP/2 on srv using a more conservative policy than
// setupHTTP2_ServeTLS because Serve is called after tls.Listen,
// and may be called concurrently. See shouldConfigureHTTP2ForServe.
//
// The tests named TestTransportAutomaticHTTP2* and
// TestConcurrentServerServe in server_test.go demonstrate some
// of the supported use cases and motivations.
func (srv *Server) setupHTTP2_Serve() error {
	srv.nextProtoOnce.Do(srv.onceSetNextProtoDefaults_Serve)
	return srv.nextProtoErr
}

func (srv *Server) onceSetNextProtoDefaults_Serve() {
	if srv.shouldConfigureHTTP2ForServe() {
		srv.onceSetNextProtoDefaults()
	}
}
``` 

`setupHTTP2_Serve` and `setupHTTP2_ServeTLS` are basically the same. `_ServeTLS` calls `onceSetNextProtoDefaults` directly, whereas `setupHTTP2_Serve` performs a check in which it confirms whether the server has a `TLSConfig` or not. In our case, it will have one. 

We can simplify this block of code with:

```
go func() {
     err = p.scadaServer.Serve(listener)
}
```

Reply:
Ah, thanks for digging into that and providing that clarification. That makes more sense.

Reply:
See https://github.com/hashicorp/vault/pull/20053/commits/a6ff6e9d2fc977fb050084328f226de55d6038b1